lib/AuthStateManager.ts

```diff
@@ -0,0 +1,129 @@
+import { AuthSdkError } from './errors';
+import PromiseQueue from './PromiseQueue';
+import { omit } from './util';
+import { 
+  ACCESS_TOKEN_STORAGE_KEY, 
+  ID_TOKEN_STORAGE_KEY 
+} from './constants';
+import { 
+  AuthState,
+  Tokens, 
+  AccessToken, 
+  IDToken,
+  UpdateAuthStateOptions
+} from './types';
+import { OktaAuth } from './browser';
+
+const DEFAULT_AUTH_STATE = { 
+  isPending: true,
+  isAuthenticated: false,
+  idToken: null,
+  accessToken: null,
+};
+const EVENT_AUTH_STATE_CHANGE = 'authStateChange';
+
+class AuthStateManager {
+  private sdk: OktaAuth;
+  private pending: { tokens: Tokens };
+  private authStateQueue: PromiseQueue;
+  private authState: AuthState;
+
+  constructor(sdk: OktaAuth) {
+    if (!sdk.emitter) {
+      throw new AuthSdkError('Emitter should be initialized before AuthStateManager');
+    }
+
+    this.sdk = sdk;
+    this.pending = { tokens: {} };
+    this.authStateQueue = new PromiseQueue();
+    this.authState = { ...DEFAULT_AUTH_STATE };
+
+    // Listen on tokenManager events to sync update tokens in memory (this.pending)
+    // And push async update authState and emit process into PromiseQ
+
+    // "added" event is emitted in both add and renew process
+    // Only listen on "added" (instead of both "added" and "renewed") to limit authState re-evaluation
+    sdk.tokenManager.on('added', (key, token) => {
+      this.pending.tokens = { ...this.pending.tokens, [key]: token };
+      this.authStateQueue.push(
+        this.updateAuthState, 
+        this,
+        { ...this.pending.tokens }
+      );
+    });
+    sdk.tokenManager.on('removed', (key) => {
+      this.pending.tokens = omit(this.pending.tokens, key);
+      this.authStateQueue.push(
+        this.updateAuthState, 
+        this,
+        { ...this.pending.tokens }
+      );
+    });
+  }
+
+  getAuthState(): AuthState {
+    return this.authState;
+  }
+
+  async initialAuthState(): Promise<void> {
+    let accessToken = await this.sdk.tokenManager.get(ACCESS_TOKEN_STORAGE_KEY) as AccessToken;
+    if (accessToken && this.sdk.tokenManager.hasExpired(accessToken)) {
+      accessToken = null;
+    }
+    let idToken = await this.sdk.tokenManager.get(ID_TOKEN_STORAGE_KEY) as IDToken;
+    if (idToken && this.sdk.tokenManager.hasExpired(idToken)) {
+      idToken = null;
+    }
+    this.updateAuthState({ accessToken, idToken });
+  }
+
+  updateAuthState({
+    accessToken, 
+    idToken,
+    shouldEvaluateIsAuthenticated = true
+  }: UpdateAuthStateOptions): Promise<void> {
+    let promise
+    if (shouldEvaluateIsAuthenticated) {
```

Comment:
Under what condition would it NOT evaluate `isAuthenticated` ?

Reply:
on "renewed" event (as it will only be emitted in success case). I removed the listener for the "renewed" event, but users still may want to build the data flow by themselves, so just kept this flag for convenience.

Reply:
Is this another optimization? Since renewed is only emitted in success case, the idea is that `isAuthenticated` should not change (should remain true.. or false?). You mentioned that you removed the listener, maybe this is because "added"/"removed" make more sense for updating auth state? (We don't really need to know if the renew *operation* succeeded or failed, but we want to track if tokens are added or removed because they are possible inputs to the `isAuthenticated` function). In general I think (per React philosophy) it is better to just blindly run the calculations the same way on every cycle and have no input parameters to the `updateAuthState` function. This makes it more rational and deterministic. Knowing that it behaves the same way every time, and that it is fire-and-forget, helps the callers. They can just call `updateAuthState` if they THINK something may have changed. If nothing changed then there will be no new authState object created and it will be a no-op.

Reply:
right, resolves to `true` is for `renewed` event, leave it without the `renewed` context can be problematic. and the optimataztion can be achieved by just providing a customized `isAuthenticated` fn. 
Will remove the flag.